File: gpt-4-turbo/stage2/loop_construct_reduction_clause.f90
!T1:loop construct reduction clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER :: i, sum, expected_sum
        INTEGER, PARAMETER :: N = 100
        
        sum = 0
        expected_sum = N*(N+1) / 2  ! Formula for the sum of the first N natural numbers

!$acc parallel loop reduction(+:sum)
        DO i = 1, N
          sum = sum + i
        END DO
!$acc end parallel

        IF (sum .ne. expected_sum) THEN
          errors = errors + 1
        END IF

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END FUNCTION test1

      PROGRAM main
        IMPLICIT NONE
        INTEGER :: failcode, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        EXTERNAL test1
        failed = .FALSE.
        failcode = 0
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM main

File: gpt-4-turbo/stage2/loop_construct_reduction_clause.c
#include <stdio.h>
#include <stdlib.h>
#include <openacc.h>

#ifndef T1
// T1: loop construct reduction clause, V:2.7-3.3
int test1() {
    int err = 0;
    const int N = 1000;
    int accumulator = 0;
    int *data = (int *)malloc(N * sizeof(int));

    // Initialize data to 1's
    for (int i = 0; i < N; ++i) {
        data[i] = 1;
    }

    #pragma acc parallel loop reduction(+:accumulator)
    for (int i = 0; i < N; ++i) {
        accumulator += data[i];
    }

    // Check if the reduction was correct
    if (accumulator != N) {
        err++;
    }

    free(data);

    return err;
}
#endif

int main() {
    const int NUM_TEST_CALLS = 10;
    int failcode = 0;
    int failed = 0;

    #ifndef T1
    for (int x = 0; x < NUM_TEST_CALLS; ++x) {
        failed = failed + test1();
    }
    if (failed != 0) {
        failcode = failcode + (1 << 0);
    }
    #endif

    return failcode;
}